pic-sure-auth-services/src/main/java/edu/harvard/hms/dbmi/avillach/auth/service/impl/OauthUserMatchingService.java

```diff
@@ -0,0 +1,129 @@
+import org.springframework.stereotype.Service;
+
+/**
+ * <p>Matches users created by admins with user profiles created by a 3rd party Oauth provider.</p>
+ */
+@Service
+public class OauthUserMatchingService {
+
+	private final Logger logger = LoggerFactory.getLogger(OauthUserMatchingService.class);
+
+	private final UserRepository userRepo;
+
+	private final UserService userService;
+
+	private final UserMetadataMappingService mappingService;
+
+	private final ConnectionRepository connectionRepo;
+
+	private final ObjectMapper mapper = new ObjectMapper();
+
+	@Autowired
+    public OauthUserMatchingService(UserRepository userRepo, UserService userService, UserMetadataMappingService mappingService, ConnectionRepository connectionRepo) {
+        this.userRepo = userRepo;
+        this.userService = userService;
+        this.mappingService = mappingService;
+        this.connectionRepo = connectionRepo;
+    }
+
+	/**
+	 * Retrieve a user profile by access_token and match it to a pre-created user in the database using
+	 * pre-configured matching rules. 
+	 * 
+	 * @param userInfo UserInfo returned from auth0
+	 * @return The user that was matched or null if no match was possible.
+	 */
+	public User matchTokenToUser(JsonNode userInfo) {
+		// This retrieves a map of UserInfo as JSON.
+		try {
+			String userInfoString = mapper.writeValueAsString(userInfo);
+            logger.info("Attempting to find match for user with info: {}", userInfo);
+
+			//Parse this once so it doesn't get re-parsed every time we read from it
+			Configuration conf = Configuration.defaultConfiguration().addOptions(Option.DEFAULT_PATH_LEAF_TO_NULL).addOptions(Option.ALWAYS_RETURN_LIST);
+			Object parsedInfo = conf.jsonProvider().parse(userInfoString);
+			//Return lists or null so that we don't have to worry about whether it's a single object or an array, or catch errors
+			List<String> connections = JsonPath.using(conf).parse(parsedInfo).read("$.identities[0].connection");
+			String connectionId = connections.getFirst();
+			Optional<Connection> connection = connectionRepo.findById(connectionId);
+			if (connection.isEmpty()) {
+				//We don't have a connection for this user
+				logger.warn("Unable to find connection with id {}", connectionId);
+				return null;
+			}
+
+			List<UserMetadataMapping> mappings = mappingService.getAllMappingsForConnection(connection.get());
+
+			if (mappings == null || mappings.isEmpty()) {
+				//We don't have any mappings for this connection yet
+                logger.warn("Unable to find user metadata mappings for connectionId {}", connection);
+				return null;
+			}
+
+			//We only care about unmatched users
+			List<User> users = userRepo.findByConnectionAndMatched(connection.orElse(null), false);
+			if (users == null || users.isEmpty()) {
+                logger.info("No unmatched users exist with connectionId {}", connection);
+				return null;
+			}
+			for (UserMetadataMapping umm : mappings) {
```

Comment:
Can't dig too deeply, but this feels like it could be optimized. I think some of the parsing could be done once instead of once per umm.